File: src/js/core/src/revocation.ts
// SPDX-License-Identifier: Apache-2.0
import forge from 'node-forge';

export interface RevocationManagerOptions {
  apiUrl: string;
  disableExternalCalls?: boolean;
  crlMemoryOnly?: boolean;
  crlCachePath?: string;
  crlRefreshIntervalHours?: number;
  debugLogging?: boolean;
}

export class RevocationManager {
  private revokedFingerprints: Set<string> = new Set();
  private lastSyncDate: string | null = null;
  private refreshInterval: any = null;

  constructor(private options: RevocationManagerOptions) {
    this.loadFromCache();
    this.startAutoRefresh();
  }

  private async loadFromCache() {
    if (this.options.crlMemoryOnly) return;
    
    // In Web/JS environment, we might use localStorage if no crlCachePath is provided
    if (typeof localStorage !== 'undefined') {
      try {
        const data = localStorage.getItem('lukuid_crl');
        if (data) {
          const json = JSON.parse(data);
          this.lastSyncDate = json.last_sync;
          if (Array.isArray(json.fingerprints)) {
            this.revokedFingerprints = new Set(json.fingerprints.map((f: string) => f.toLowerCase()));
          }
        }
      } catch (e) {
        this.debugLog('Failed to load CRL cache from localStorage', e);
      }
    }
  }

  private saveToCache() {
    if (this.options.crlMemoryOnly) return;

    if (typeof localStorage !== 'undefined') {
      try {
        const json = {
          last_sync: this.lastSyncDate,
          fingerprints: Array.from(this.revokedFingerprints)
        };
        localStorage.setItem('lukuid_crl', JSON.stringify(json));
      } catch (e) {
        this.debugLog('Failed to save CRL cache to localStorage', e);
      }
    }
  }

  private startAutoRefresh() {
    if (!this.options.crlRefreshIntervalHours || this.options.crlRefreshIntervalHours <= 0) return;

    const intervalMs = this.options.crlRefreshIntervalHours * 3600 * 1000;
    this.refreshInterval = setInterval(() => this.sync(), intervalMs);
    
    // Initial sync
    this.sync().catch(e => this.debugLog('Initial CRL sync failed', e));
  }

  public async sync() {
    if (this.options.disableExternalCalls) return;

    const apiUrl = this.options.apiUrl.replace(/\/+$/, '');
    let url = `${apiUrl}/revocations`;
    if (this.lastSyncDate) {
      url += `?from=${encodeURIComponent(this.lastSyncDate)}`;
    }

    try {
      const response = await fetch(url);
      if (response.status === 200) {
        const json = await response.json();
        if (Array.isArray(json.revocations)) {
          for (const item of json.revocations) {
            if (item.identifier) {
              this.revokedFingerprints.add(item.identifier.toLowerCase());
            }
          }
        }
        this.lastSyncDate = new Date().toISOString();
        this.saveToCache();
        this.debugLog(`CRL synced successfully: ${this.revokedFingerprints.size} items`);
      } else if (response.status !== 304) {
        this.debugLog(`CRL sync failed with status: ${response.status}`);
      }
    } catch (e) {
      this.debugLog('CRL sync error', e);
    }
  }

  public isRevoked(cert: any): boolean {
    const fingerprint = this.getFingerprint(cert);
    return this.revokedFingerprints.has(fingerprint.toLowerCase());
  }

  private getFingerprint(cert: any): String {
    // For Ed25519, we need to extract the raw bytes.
    // forge might not support Ed25519 natively in a way that is easy to extract.
    // In LukuID JS SDK, we often use noble-forge or similar.
    
    // Fallback: extract public key bytes from SPKI
    const f = forge as any;
    try {
        const spkiDer = f.asn1.toDer(f.pki.publicKeyToAsn1(cert.publicKey)).getBytes();
        const asn1 = f.asn1.fromDer(spkiDer);
        // SPKI: SEQUENCE { algorithm SEQUENCE { oid OID, params ANY }, publicKey BIT STRING }
        const publicKeyBitString = asn1.value[1].value;
        const rawBytes = publicKeyBitString.slice(1); // Drop unused bits byte
        
        const md = f.md.sha256.create();
        md.update(rawBytes);
        return md.digest().toHex();
    } catch (e) {
        // Fallback to full cert hash
        const der = f.asn1.toDer(f.pki.certificateToAsn1(cert)).getBytes();
        const md = f.md.sha256.create();
        md.update(der);
        return md.digest().toHex();
    }
  }

  private debugLog(message: string, error?: any) {
    if (this.options.debugLogging) {
      console.debug(`[RevocationManager] ${message}`, error || '');
    }
  }

  public close() {
    if (this.refreshInterval) {
      clearInterval(this.refreshInterval);
      this.refreshInterval = null;
    }
  }
}

File: src/python/lukuid_sdk/revocation.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import json
import os
import threading
import time
from datetime import datetime, timezone
from typing import Set, Any
import requests
from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization

class RevocationManager:
    def __init__(self, options: dict[str, Any]):
        self.options = options
        self.revoked_fingerprints: Set[str] = set()
        self.last_sync_date: str | None = None
        self._lock = threading.Lock()
        self._stop_event = threading.Event()
        self._refresh_thread: threading.Thread | None = None

        self.load_from_cache()
        self.start_auto_refresh()

    def _get_cache_file(self) -> str | None:
        if self.options.get("crl_memory_only"):
            return None
        
        cache_path = self.options.get("crl_cache_path")
        if not cache_path:
            # Default to user's home directory or current dir
            cache_path = os.path.expanduser("~/.lukuid")
        
        if not os.path.exists(cache_path):
            try:
                os.makedirs(cache_path, exist_ok=True)
            except Exception:
                return None
        
        return os.path.join(cache_path, "lukuid_crl.json")

    def load_from_cache(self):
        file_path = self._get_cache_file()
        if not file_path or not os.path.exists(file_path):
            return

        try:
            with open(file_path, "r") as f:
                data = json.load(f)
                self.last_sync_date = data.get("last_sync")
                fingerprints = data.get("fingerprints", [])
                with self._lock:
                    self.revoked_fingerprints = set(f.lower() for f in fingerprints)
        except Exception:
            pass

    def save_to_cache(self):
        file_path = self._get_cache_file()
        if not file_path:
            return

        try:
            data = {
                "last_sync": self.last_sync_date,
                "fingerprints": list(self.revoked_fingerprints)
            }
            with open(file_path, "w") as f:
                json.dump(data, f)
        except Exception:
            pass

    def start_auto_refresh(self):
        interval_hours = self.options.get("crl_refresh_interval_hours", 4)
        if interval_hours <= 0:
            return

        def refresh_loop():
            while not self._stop_event.is_set():
                try:
                    self.sync()
                except Exception:
                    pass
                self._stop_event.wait(interval_hours * 3600)

        self._refresh_thread = threading.Thread(target=refresh_loop, daemon=True)
        self._refresh_thread.start()

    def sync(self):
        if self.options.get("disable_external_calls"):
            return

        api_url = self.options.get("api_url", "https://api.lukuid.com").rstrip("/")
        url = f"{api_url}/revocations"
        if self.last_sync_date:
            url += f"?from={self.last_sync_date}"

        try:
            response = requests.get(url, timeout=10)
            if response.status_code == 200:
                data = response.json()
                revocations = data.get("revocations", [])
                with self._lock:
                    for item in revocations:
                        fingerprint = item.get("identifier")
                        if fingerprint:
                            self.revoked_fingerprints.add(fingerprint.lower())
                
                self.last_sync_date = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
                self.save_to_cache()
            elif response.status_code != 304:
                pass
        except Exception:
            pass

    def is_revoked(self, cert: x509.Certificate) -> bool:
        fingerprint = self.get_fingerprint(cert)
        with self._lock:
            return fingerprint.lower() in self.revoked_fingerprints

    def get_fingerprint(self, cert: x509.Certificate) -> str:
        # Extract raw public key bytes
        public_key = cert.public_key()
        if isinstance(public_key, ed25519.Ed25519PublicKey):
            raw_bytes = public_key.public_bytes(
                encoding=serialization.Encoding.Raw,
                format=serialization.PublicFormat.Raw
            )
        else:
            # For others, extract from SubjectPublicKeyInfo bit string
            # In cryptography, we can use public_bytes(DER, SubjectPublicKeyInfo)
            # but that's the whole SPKI.
            # To get consistent with others, we'll try to get the raw bytes.
            if isinstance(public_key, ec.EllipticCurvePublicKey):
                raw_bytes = public_key.public_bytes(
                    encoding=serialization.Encoding.X962,
                    format=serialization.PublicFormat.UncompressedPoint
                )
            else:
                # Fallback to full cert DER
                raw_bytes = cert.public_bytes(serialization.Encoding.DER)
        
        digest = hashlib.sha256(raw_bytes).hexdigest()
        return digest.lower()

    def close(self):
        self._stop_event.set()
        if self._refresh_thread:
            # We don't join because it's a daemon thread and might be sleeping for hours
            pass
